lookup_optim.go

```diff
@@ -0,0 +1,306 @@
+	}, nil
+}
+
+func (dht *IpfsDHT) optimisticProvide(outerCtx context.Context, keyMH multihash.Multihash) error {
+	key := string(keyMH)
+
+	if key == "" {
+		return fmt.Errorf("can't lookup empty key")
+	}
+
+	// initialize new context for all putProvider operations.
+	// We don't want to give the outer context to the put operations as we return early before all
+	// put operations have finished to avoid the long tail of the latency distribution. If we
+	// provided the outer context the put operations may be cancelled depending on what happens
+	// with the context on the user side.
+	putCtx, putCtxCancel := context.WithTimeout(context.Background(), time.Minute)
+
+	es, err := dht.newOptimisticState(putCtx, key)
+	if err != nil {
+		putCtxCancel()
+		return err
+	}
+
+	// initialize context that finishes when this function returns
+	innerCtx, innerCtxCancel := context.WithCancel(outerCtx)
+	defer innerCtxCancel()
+
+	go func() {
+		select {
+		case <-outerCtx.Done():
+			// If the outer context gets cancelled while we're still in this function. We stop all
+			// pending put operations.
+			putCtxCancel()
+		case <-innerCtx.Done():
+			// We have returned from this function. Ignore cancellations of the outer context and continue
+			// with the remaining put operations.
+		}
+	}()
+
+	lookupRes, err := dht.runLookupWithFollowup(outerCtx, key, dht.pmGetClosestPeers(key), es.stopFn)
+	if err != nil {
+		return err
+	}
+
+	// Store the provider records with all the closest peers
+	// we haven't already contacted.
+	es.peerStatesLk.Lock()
+	for _, p := range lookupRes.peers {
+		if _, found := es.peerStates[p]; found {
+			continue
+		}
+
+		go es.putProviderRecord(p)
+		es.peerStates[p] = Sent
+	}
+	es.peerStatesLk.Unlock()
+
+	// wait until a threshold number of RPCs have completed
+	es.waitForRPCs()
+
+	if err := outerCtx.Err(); err != nil || !lookupRes.completed { // likely the "completed" field is false but that's not a given
+		return err
+	}
+
+	// tracking lookup results for network size estimator as "completed" is true
+	if err = dht.nsEstimator.Track(key, lookupRes.closest); err != nil {
+		logger.Warnf("network size estimator track peers: %s", err)
+	}
+
+	// refresh the cpl for this key as the query was successful
+	dht.routingTable.ResetCplRefreshedAtForID(kb.ConvertKey(key), time.Now())
+
+	return nil
+}
+
+func (os *optimisticState) stopFn(qps *qpeerset.QueryPeerset) bool {
+	os.peerStatesLk.Lock()
+	defer os.peerStatesLk.Unlock()
+
+	// get currently known closest peers and check if any of them is already very close.
+	// If so -> store provider records straight away.
+	closest := qps.GetClosestNInStates(os.dht.bucketSize, qpeerset.PeerHeard, qpeerset.PeerWaiting, qpeerset.PeerQueried)
+	distances := make([]float64, os.dht.bucketSize)
+	for i, p := range closest {
+		// calculate distance of peer p to the target key
+		distances[i] = netsize.NormedDistance(p, os.ksKey)
+
+		// Check if we have already interacted with that peer
```

Comment:
I think the wording can be more precise here, we haven't necessarily interacted with the peer when it's in `peerStates`, it just means that we've scheduled interaction with the peer.  (There are some other comments with this same ambiguity.)

Reply:
Just to understand you correctly. Do you mean the brief moment of time when we have stored e.g. `Sent` into `peerStates` until the corresponding goroutine `putProviderRecord` was scheduled? This would indeed be an instance where we haven't interacted with the peer but just scheduled interaction 👍

Reply:
yeah that's all I meant, it's definitely a nit but just trying to be precise